src/dns_server.rs

```diff
@@ -0,0 +1,187 @@
+use log::{info};
+use tokio;
+use tokio::net::{UdpSocket, ToSocketAddrs};
+use tokio::sync::Mutex;
+use trust_dns_proto::op::{Message};
+use trust_dns_proto::rr::{Record, RecordType, DNSClass, RData, Name};
+use std::net::{Ipv4Addr};
+use trust_dns_proto::serialize::binary::BinEncodable;
+use std::sync::Arc;
+
+use crate::mapping::Ipv4Mapping;
+
+
+const BASE_IPV4_ADDR: u32 = 0x0D000000;
+const FIRST_IPV4_ADDR: u32 = BASE_IPV4_ADDR + 1;
+
+#[derive(Debug)]
+pub struct DnsServer {
+    dns_server: UdpSocket,
+    answer_index: u32,
+    ipv4_mapping: Arc<Mutex<Ipv4Mapping>>,
+}
+
+impl DnsServer {
+    pub async fn bind<A: ToSocketAddrs>(addr: A) -> Result<DnsServer, tokio::io::Error> {
+        let server = DnsServer {
+            dns_server: UdpSocket::bind(addr).await?,
+            answer_index: 1,
+            ipv4_mapping: Arc::new(Mutex::new(Ipv4Mapping::default())),
+        };
+        Ok(server)
+    }
+
+    pub fn port(&self) -> Result<u16, tokio::io::Error> {
+        self.dns_server.local_addr().map(|addr| addr.port())
+    }
+
+    pub fn ipv4_mapping(&self) -> Arc<Mutex<Ipv4Mapping>> {
+        self.ipv4_mapping.clone()
+    }
+
+    pub async fn run(mut self) -> Result<(), tokio::io::Error> {
+        let mut query_buf = vec![0; 65536];
+        loop {
+            let (_, addr) = self.dns_server.recv_from(&mut query_buf).await?;
+            let mut msg = match Message::from_vec(&query_buf) {
+                Ok(query) => query,
+                Err(_) => {
+                    info!("Received invalid dns packet");
+                    continue;
+                }
+            };
+            let mut answers = Vec::with_capacity(msg.query_count() as usize);
+            for query in msg.queries() {
+                let mut record = Record::new();
+                record.set_name(query.name().clone());
+                record.set_record_type(RecordType::A);
+                record.set_dns_class(DNSClass::IN);
+                record.set_ttl(60 * 60);  // 1 hour
```

Comment:
const hour_in_seconds: u32 = 3600;

Reply:
In what way it will make it better? it's used once, here, and there's a comment that says the meaning of it

Reply:
Classic Oded comment isn't it? 
I like it this way, but what ever